File: core/src/main/java/com/github/jarol/azure/search/spark/sql/connector/core/schema/conversion/output/AtomicDecoders.java
package com.github.jarol.azure.search.spark.sql.connector.core.schema.conversion.output;

import com.github.jarol.azure.search.spark.sql.connector.core.utils.StringUtils;
import com.github.jarol.azure.search.spark.sql.connector.core.utils.TimeUtils;
import org.apache.spark.unsafe.types.UTF8String;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.time.format.DateTimeFormatter;

/**
 * Collection of atomic {@link SearchDecoder}
 */

public final class AtomicDecoders {

    /**
     * Gets a no-op decoder (i.e. a decoder that does not apply any transformation)
     * @return a no-op decoder
     */

    @Contract(pure = true)
    public static @NotNull SearchDecoder identity() {

        return value -> value;
    }

    /**
     * Get the decoder for strings
     * @return decoder for strings
     */

    @Contract(value = " -> new", pure = true)
    public static @NotNull SearchDecoder forUTF8Strings() {

        // Strings are internally represented as UTF8Strings
        return new TransformDecoder<String>() {
            @Override
            protected String transform(Object value) {
                return StringUtils.fromUTF8String(
                        (UTF8String) value
                );
            }
        };
    }

    /**
     * Gets a decoder from numeric/boolean types to string
     * @return a decoder from numeric/boolean types to string
     */

    @Contract(value = " -> new", pure = true)
    public static @NotNull SearchDecoder stringValueOf() {

        return new TransformDecoder<String>() {
            @Override
            protected String transform(Object value) {
                return String.valueOf(value);
            }
        };
    }

    /**
     * Gets a decoder for dates
     * @return decoder for dates
     */

    @Contract(value = " -> new", pure = true)
    public static @NotNull SearchDecoder forDates() {

        return new TimeDecoder() {
            @Override
            protected OffsetDateTime toOffsetDateTime(Object value) {

                return TimeUtils.fromEpochDays((Integer) value);
            }
        };
    }

    /**
     * Gets a decoder from dates to strings
     * @return decoder from dates to strings
     */

    @Contract(value = " -> new", pure = true)
    public static @NotNull SearchDecoder fromDateToString() {

        return new TransformDecoder<String>() {

            @Override
            protected String transform(Object value) {
                return LocalDate
                        .ofEpochDay((Integer) value)
                        .format(DateTimeFormatter.ISO_LOCAL_DATE);
            }
        };
    }

    /**
     * Gets a decoder for timestamps
     * @return decoder for timestamps
     */

    @Contract(value = " -> new", pure = true)
    public static @NotNull SearchDecoder forTimestamps() {

        return new TimeDecoder() {
            @Override
            protected OffsetDateTime toOffsetDateTime(Object value) {
                return TimeUtils.fromEpochMicros((Long) value);
            }
        };
    }
}

File: core/src/main/scala/com/github/jarol/azure/search/spark/sql/connector/core/schema/conversion/output/AtomicDecoders.scala
package com.github.jarol.azure.search.spark.sql.connector.core.schema.conversion.output

import com.github.jarol.azure.search.spark.sql.connector.core.Constants
import org.apache.spark.unsafe.types.UTF8String

import java.nio.charset.StandardCharsets
import java.time.temporal.ChronoUnit
import java.time.{Instant, LocalDate, LocalTime, OffsetDateTime}

/**
 * Converter that applies a transformation to a non-null Spark internal value
 * @tparam T Search target type
 */

trait TransformDecoder[T]
  extends SearchDecoder {

  override final def apply(value: Any): T = {

    Option(value)
      .map(transform)
      .getOrElse(null.asInstanceOf[T])
  }

  /**
   * Apply the transformation
   *
   * @param value Spark internal value
   * @return a Search document property
   */

  protected def transform(value: Any): T
}

/**
 * Converter from Spark internal time types to Search datetime type
 */

sealed trait TimeDecoder
  extends TransformDecoder[String] {

  override final protected def transform(value: Any): String = {

    // Convert the Spark internal object to date time and then format it as a string
    toOffsetDateTime(value).format(
      Constants.DATETIME_OFFSET_FORMATTER
    )
  }

  /**
   * Transform the Spark internal value to an [[OffsetDateTime]]
   * @param value Spark internal value
   * @return an offset date time
   */

  protected def toOffsetDateTime(value: Any): OffsetDateTime
}


object AtomicDecoders {

  final val STRING_VALUE_OF: TransformDecoder[String] = (value: Any) => String.valueOf(value)

  final val STRING: TransformDecoder[String] = (value: Any) => new String(value.asInstanceOf[UTF8String].getBytes, StandardCharsets.UTF_8)

  /**
   * Converter for dates
   */

  final val DATE: TimeDecoder = new TimeDecoder {
    override protected def toOffsetDateTime(value: Any): OffsetDateTime = {
      OffsetDateTime.of(
        LocalDate.ofEpochDay(value.asInstanceOf[Int].toLong),
        LocalTime.MIDNIGHT,
        Constants.UTC_OFFSET
      )
    }
  }

  /**
   * Converter for timestamps
   */

  final val TIMESTAMP: TimeDecoder = new TimeDecoder {
    override protected def toOffsetDateTime(value: Any): OffsetDateTime = {

      Instant.EPOCH.plus(
        value.asInstanceOf[Long],
        ChronoUnit.MICROS
      ).atOffset(Constants.UTC_OFFSET)
    }
  }
}
